### PR TITLE
exp/services/ledgerexporter: Add prometheus metrics for ledger exporter

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://developers.stellar.org/api/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Adds prometheus metrics to the ledger exporter service. The metrics are served via HTTP through an admin port which is specified in the ledger exporter configuration. Some notable metrics included:

- latest ledger sequence consumed by the exporter
- upload duration
- tx meta object size (both compressed and uncompressed)
- number of pending upload tasks

### Why

https://stellarorg.atlassian.net/browse/HUBBLE-277

### Known limitations

[N/A]

### CODE REVIEW

File: exp/services/ledgerexporter/internal/app.go [hunk 1]
```diff
@@ -62,17 +95,19 @@ func (a *App) Run() {
...
-
-		err := a.uploader.Run(ctx)
-		if err != nil && !errors.Is(err, context.Canceled) {
-			logger.WithError(err).Error("Error executing Uploader")
-			cancel()
-		}
-	}()
+	wg.Add(1 + a.config.UploadWorkers)
+
+	for i := 0; i < a.config.UploadWorkers; i++ {
```
Comment (sreuland) [inline exp/services/ledgerexporter/internal/app.go:100]: this wasn't required for metrics correct, is just some tech debt on the side?

Comment (tamirms) [reply exp/services/ledgerexporter/internal/app.go:100]: correct, this change is a side effect of making the queue size configurable. I refactored the queue into its own type to make it easier to instrument metrics and that lead to having upload workers configurable

File: exp/services/ledgerexporter/internal/exportmanager.go [hunk 2]
```diff
@@ -14,35 +17,32 @@ type ExporterConfig struct {
 }
 
 // ExportManager manages the creation and handling of export objects.
-type ExportManager interface {
```
Comment (sreuland) [inline exp/services/ledgerexporter/internal/exportmanager.go:17]: why get rid of the Interface? there was some consistency with Managers expressed as types rather than structs, also, it helps for ongoing unit tests, to be able to mock out.

File: exp/services/ledgerexporter/internal/queue.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
+package ledgerexporter
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// UploadQueue is a queue of LedgerMetaArchive objects which are scheduled for upload
+type UploadQueue struct {
```
Comment (sreuland) [inline exp/services/ledgerexporter/internal/queue.go:10]: this is nice addition, but not sure why is it being added as feature scope to this pr's intent of adding metrics to the existing app, the concerns seem unrelated. I noticed https://stellarorg.atlassian.net/browse/HUBBLE-344 which seem to be the applicable ticket for this parallelization of uploaders enhancement to be done under?

Comment (tamirms) [reply exp/services/ledgerexporter/internal/queue.go:10]: I wanted to expose a metric for the size of the queue and it was easier to do that after refactoring the queue to be its own component

File: exp/services/ledgerexporter/internal/exportmanager.go (hunk 2 above)
Comment (tamirms) [reply exp/services/ledgerexporter/internal/exportmanager.go:17]: the export manager interface was not used. 

"Do not define interfaces before they are used" is a recommendation found in the golang style guide https://go.dev/wiki/CodeReviewComments#interfaces

File: exp/services/ledgerexporter/internal/queue.go (hunk 3 above)
Comment (tamirms) [reply exp/services/ledgerexporter/internal/queue.go:10]: the UploadQueue refactoring still does address https://stellarorg.atlassian.net/browse/HUBBLE-344 because I have not done any sort of benchmarking to asses whether having multiple upload workers will be effective when reingesting large ranges. IMO, https://stellarorg.atlassian.net/browse/HUBBLE-344 is still valid even after merging this pr

File: exp/services/ledgerexporter/internal/uploader.go [hunk 4]
```diff
@@ -2,47 +2,116 @@ package ledgerexporter
...
+		},
+		[]string{"already_exists", "ledgers"},
+	)
+	objectSizeMetrics := prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace: "ledger_exporter", Subsystem: "uploader", Name: "object_size_bytes",
+			Help:       "size of a ledger batch in bytes, sliding window = 10m",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"ledgers", "already_exists", "compression"},
```
Comment (urvisavla) [inline exp/services/ledgerexporter/internal/uploader.go:41]: `already_exists` doesn't seem  to be relevant for `objectSizeMetrics`?

File: exp/services/ledgerexporter/internal/uploader.go [hunk 5]
```diff
@@ -2,47 +2,116 @@ package ledgerexporter
...
+	)
+	prometheusRegistry.MustRegister(uploadDurationMetric)
```
Comment (urvisavla) [inline exp/services/ledgerexporter/internal/uploader.go:43]: `objectSizeMetrics` need to be registered as well.

File: exp/services/ledgerexporter/internal/queue.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	prometheusRegistry.MustRegister(queueLengthMetric)
+	return UploadQueue{
+		metaArchiveCh:     make(chan *LedgerMetaArchive, size),
+		queueLengthMetric: queueLengthMetric,
+	}
+}
+
+// Enqueue will add an upload task to the queue. Enqueue may block if the queue is full.
+func (u UploadQueue) Enqueue(ctx context.Context, archive *LedgerMetaArchive) error {
+	u.queueLengthMetric.Inc()
```
Comment (urvisavla) [inline exp/services/ledgerexporter/internal/queue.go:32]: Should we increment only after the object has been added to the queue on line 34? This is to account queue is full case and adding to the queue is blocked, similar to how it's handled in Dequeue.

File: exp/services/ledgerexporter/internal/uploader.go [hunk 7]
```diff
@@ -2,47 +2,116 @@ package ledgerexporter
...
+		count:  &r.totalCompressed,
+	})
+	r.totalUncompressed += uncompressedCount
+	return uncompressedCount, err
 }
 
 // Upload uploads the serialized binary data of ledger TxMeta to the specified destination.
-// TODO: Add retry logic.
-func (u *uploader) Upload(ctx context.Context, metaArchive *LedgerMetaArchive) error {
+func (u Uploader) Upload(ctx context.Context, metaArchive *LedgerMetaArchive) error {
```
Comment (urvisavla) [inline exp/services/ledgerexporter/internal/uploader.go:79]: In addition to the `export_manager_latest_ledger` metric for the last ingested ledger in the export manager it might also be useful to have a metric for the last uploaded ledger sequence number here in the uploader?

File: exp/services/ledgerexporter/internal/uploader.go (hunk 4 above)
Comment (tamirms) [reply exp/services/ledgerexporter/internal/uploader.go:41]: if the file already exists the request may not need to upload the entire file to GCS because GCS can detect if the file already exists early on. So this tag is necessary to determine if we are spending full bandwidth to upload files that already exist

File: exp/services/ledgerexporter/internal/queue.go (hunk 6 above)
Comment (tamirms) [reply exp/services/ledgerexporter/internal/queue.go:32]: Yeah, this behavior is actually intentional.

If the channel is full there is still an implicit queue of excess archives that are blocked in this function trying to push to the channel. It is good that this metric will still give us visibility into that case.

File: exp/services/ledgerexporter/internal/uploader.go (hunk 7 above)
Comment (tamirms) [reply exp/services/ledgerexporter/internal/uploader.go:79]: I didn't want to implement that in case there were multiple upload workers. With multiple upload workers, each worker would have a different latest ledger. There are different ways to resolve that issue but I think it's better to do it as a part of https://stellarorg.atlassian.net/browse/HUBBLE-344

File: exp/services/ledgerexporter/internal/app.go (hunk 1 above)
Comment (tamirms) [reply exp/services/ledgerexporter/internal/app.go]: I removed the upload workers configuration and hard coded the queue size to 1

File: exp/services/ledgerexporter/internal/queue.go (hunk 3 above)
Comment (tamirms) [reply exp/services/ledgerexporter/internal/queue.go:10]: I removed the upload workers configuration and hard coded the queue size to 1

**Comment by tamirms** [comment]
@sreuland @urvisavla PTAL I believe all the feedback has been addressed

File: exp/services/ledgerexporter/internal/uploader.go (hunk 7 above)
Comment (urvisavla) [reply exp/services/ledgerexporter/internal/uploader.go:79]: That's a valid point. Another option would be to add the metric now assuming a single upload worker (as currently implemented) and handle the multiple upload workers in https://stellarorg.atlassian.net/browse/HUBBLE-344.

File: exp/services/ledgerexporter/internal/app.go [hunk 8]
```diff
@@ -97,20 +133,23 @@ func (a *App) Run() {
...
+func mustNewLedgerBackend(ctx context.Context, config Config, prometheusRegistry *prometheus.Registry) ledgerbackend.LedgerBackend {
 	captiveConfig := config.GenerateCaptiveCoreConfig()
 
+	var backend ledgerbackend.LedgerBackend
+	var err error
 	// Create a new captive core backend
-	backend, err := ledgerbackend.NewCaptive(captiveConfig)
+	backend, err = ledgerbackend.NewCaptive(captiveConfig)
 	logFatalIf(err, "Failed to create captive-core instance")
+	backend = ledgerbackend.WithMetrics(backend, prometheusRegistry, "ledger_exporter")
```
Comment (urvisavla) [inline exp/services/ledgerexporter/internal/app.go:152]: I'm not sure if this is expected or its an issue here or in captive-core, but when I ran ledger exporter, the ledger_exporter_ingest_captive_stellar_core_supported_protocol_version always shows as -1.

```
# HELP ledger_exporter_ingest_captive_stellar_core_supported_protocol_version determines the supported version of the protocol by Captive-Core
# TYPE ledger_exporter_ingest_captive_stellar_core_supported_protocol_version gauge
ledger_exporter_ingest_captive_stellar_core_supported_protocol_version -1

```

**Review by urvisavla — APPROVED** [review]
LGTM! Just added a couple of notes for consideration.

File: exp/services/ledgerexporter/internal/uploader.go (hunk 7 above)
Comment (tamirms) [reply exp/services/ledgerexporter/internal/uploader.go:79]: I think the metric definition for the most recently uploaded ledger will change in https://stellarorg.atlassian.net/browse/HUBBLE-344 because it will likely require a metric label for the thread / worker id.

File: exp/services/ledgerexporter/internal/app.go (hunk 8 above)
Comment (tamirms) [reply exp/services/ledgerexporter/internal/app.go:152]: the version is set to -1 when we cannot probe the captive core port:

https://github.com/stellar/go/blob/master/ingest/ledgerbackend/captive_core_backend.go#L241-L243

It looks like this is occurring because we do not configure the captive core http port:

https://github.com/stellar/go/blob/master/exp/services/ledgerexporter/internal/config.go#L182-L191

We should configure the http port as a part of https://stellarorg.atlassian.net/browse/HUBBLE-292 or https://stellarorg.atlassian.net/browse/HUBBLE-271
